pkg/tpm/tcpa_log.go

```diff
@@ -0,0 +1,730 @@
+		return nil, err
+	}
+
+	guid := fmt.Sprintf("Variable - %x-%x-%x-%x-%x - ", variableData.variableName.blockA, variableData.variableName.blockB, variableData.variableName.blockC, variableData.variableName.blockD, variableData.variableName.blockE)
+	eventInfo := guid
+	utf16String := utf16.Decode(variableData.unicodeName)
+	eventInfo += fmt.Sprintf("%s", string(utf16String))
+
+	return &eventInfo, nil
+}
+
+func getEventDataString(eventType uint32, eventData []byte) (*string, error) {
+	if eventType < uint32(EvEFIEventBase) {
+		switch BIOSLogID(eventType) {
+		case EvSeparator:
+			eventInfo := fmt.Sprintf("%x", eventData)
+			return &eventInfo, nil
+		case EvAction:
+			eventInfo := string(bytes.Trim(eventData, "\x00"))
+			return &eventInfo, nil
+		case EvOmitBootDeviceEvents:
+			eventInfo := string("BOOT ATTEMPTS OMITTED")
+			return &eventInfo, nil
+		case EvPostCode:
+			eventInfo := string(bytes.Trim(eventData, "\x00"))
+			return &eventInfo, nil
+		case EvEventTag:
+			eventInfo, err := getTaggedEvent(eventData)
+			if err != nil {
+				return nil, err
+			}
+			return eventInfo, nil
+		case EvSCRTMContents:
+			eventInfo := string(bytes.Trim(eventData, "\x00"))
+			return &eventInfo, nil
+		case EvIPL:
+			eventInfo := string(bytes.Trim(eventData, "\x00"))
+			return &eventInfo, nil
+		}
+	} else {
+		switch EFILogID(eventType) {
+		case EvEFIHCRTMEvent:
+			eventInfo := "HCRTM"
+			return &eventInfo, nil
+		case EvEFIAction:
+			eventInfo := string(bytes.Trim(eventData, "\x00"))
+			return &eventInfo, nil
+		case EvEFIVariableDriverConfig, EvEFIVariableBoot, EvEFIVariableAuthority:
+			eventInfo, err := getVariableDataString(eventData)
+			if err != nil {
+				return nil, err
+			}
+			return eventInfo, nil
+		case EvEFIRuntimeServicesDriver, EvEFIBootServicesDriver, EvEFIBootServicesApplication:
+			eventInfo, err := getImageLoadEventString(eventData)
+			if err != nil {
+				return nil, err
+			}
+			return eventInfo, nil
+		case EvEFIGPTEvent:
+			eventInfo, err := getGPTEventString(eventData)
+			if err != nil {
+				return nil, err
+			}
+			return eventInfo, nil
+		case EvEFIPlatformFirmwareBlob:
+			eventInfo, err := getPlatformFirmwareBlob(eventData)
+			if err != nil {
+				return nil, err
+			}
+			return eventInfo, nil
+		case EvEFIHandoffTables:
+			eventInfo, err := getHandoffTablePointers(eventData)
+			if err != nil {
+				return nil, err
+			}
+			return eventInfo, nil
+		}
+	}
+
+	eventInfo := string(bytes.Trim(eventData, "\x00"))
+	return &eventInfo, errors.New("Event type couldn't get parsed")
+}
+
+func readTPM2Log(firmware string) (*PCRLog, error) {
+	var pcrLog PCRLog
+	pcrLog.Firmware = firmware
+
+	file, err := os.Open(DefaultTCPABinaryLog)
+	if err != nil {
+		return nil, err
+	}
+
+	var endianess binary.ByteOrder = binary.LittleEndian
+	var pcrDigest PCRDigestInfo
+	var pcrEvent TcgPcrEvent2
+	for {
+		if err := binary.Read(file, endianess, &pcrEvent.pcrIndex); err == io.EOF {
```

Comment:
read this: https://blog.golang.org/errors-are-values

then look at the Unmarshal example for Lexer here: https://godoc.org/github.com/u-root/u-root/pkg/uio#Lexer

all of these 4 line boiler plate things can be one liners, e.g. with Lexer's ReadData which uses binary.Read: https://godoc.org/github.com/u-root/u-root/pkg/uio#Lexer.ReadData

You'd have to read in the full file or build something like a LexerReader that reads on demand.

Reply:
that not only reduces boiler plate, but also human mistakes